program-analysis/echidna/Exercise-X.md

```diff
@@ -0,0 +1,442 @@
+ 40 |     |             // actions
+ 41 | r   |             naughtCoin.transfer(bob, 100);
+ 42 |     |         }
+ 43 |     |         // post-conditions
+ 44 |     |         assert(
+ 45 |     |             naughtCoin.balanceOf(player) == playerBalanceBefore &&
+ 46 |     |                 naughtCoin.balanceOf(bob) == bobBalanceBefore
+ 47 |     |         );
+ 48 |     |     }
+ 49 |     | }
+   ```
+
+   You can see that the transfer reverted as expected. But... Our post-conditions weren't checked. If you have read the [ERC20 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md), you will know that the `transfer` function returns a status `boolean`. In the case of success, we can emit a special `AssertionFailed` event that won't stop the execution of the function. This way, our post-conditions will be checked.
+
+- If `transfer` failed, that's okay. Keep going.
+- If `transfer` succeeded, we don't want that, emit the `AssertionFailed` and check post-conditions as usual.
+ 
+ Add the `event AssertionFailed(uint256 amount);` at the top of your contract.
+ 
+   ```
+   function transfer_should_fail_before_timelock_period(uint256 amount)
+        public
+    {
+        // pre-conditions
+        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+        uint256 currentTime = block.timestamp;
+        if (currentTime < naughtCoin.timeLock()) {
+            // actions
+            bool success1 = naughtCoin.transfer(bob, amount);
+            if (success1) {
+                emit AssertionFailed(amount);
+            }
+        }
+        // post-conditions
+        assert(
+            naughtCoin.balanceOf(player) == playerBalanceBefore &&
+                naughtCoin.balanceOf(bob) == bobBalanceBefore
+        );
+    }
+
+   ```
+
+Run Echidna and check the corpus. Now we have fully covered this property.
+
+</details>
+
+<details>
+  <summary>4th property</summary>
+  
+  According to the ERC20 spec, the `approve` function should not fail if the caller has enough tokens to make the approval.
+  
+   ```
+   function approve_should_not_fail_if_caller_has_enough_tokens(uint256 amount)
+        public
+    {
+        // pre-conditions
+        uint256 playerBalance = naughtCoin.balanceOf(player);
+        if (playerBalance >= amount) {
+            // actions
+            bool success1 = naughtCoin.approve(bob, amount);
+            // post-conditions
+            assert(success1);
+        }
+    }
+   ```
+
+</details>
+
+<details>
+  <summary>5th property</summary>
+  
+  Player should not `burn` tokens before the `timeLock` period.
+  What is token burning?
+  
+  You can burn tokens by sending them to the `0` address or by sending them to a non existent address. Sending to the `0` address is not possible because of the `0` address checks in the `ERC20` OZ standard. Sending them to any address shouldn't be possible as well because of the `timeLock`.
```

Comment:
This is a bit of a technicality, but transferring it to anywhere doesn't burn them, it just locks them.
A "proper" burning should decrease the total supply of the token, and not all tokens implement public functions to do so. Base [ERC20 OZ token](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/a70ee4e3bbee565e49217c9ec59de5354f8af9d8/contracts/token/ERC20/ERC20.sol#L277-L288) implements an internal _burn method that can be used by token implementations.

Reply:
This is a fair point. I think the word "effectively" is missing. Sending tokens to a non-existing address locks them up, which effectively burns them. 

The `ERC20` used in this Ethernaut challenge did not implement a public burn functionality.

Reply:
Fixed